File: Python - Contestants/ExampleSimulation.py
import matplotlib.pyplot as plt
import numpy as np
import scipy.io as sio


from Controller import Controller
from TanksMassBalance import TanksMassBalance

## Load example, pre-allocate vectors
data = sio.loadmat("Example1")

ABW = data["ABW"][:, 0]
AMW = data["AMW"][:, 0]
GUreq = data["GUreq"][:, 0]
HUreq = data["HUreq"][:, 0]
RW = data["RW"][:, 0]
t = data["t"][:, 0]

N = 365 * 24

L = np.zeros(N + 1)
MWspec = np.zeros(N)
BWspec = np.zeros(N)
MW = np.zeros(N)
BW = np.zeros(N)
HU = np.zeros(N)
GU = np.zeros(N)
OF = np.zeros(N)

## Initialize model
# Tank is halfway full
L[:2] = 2.155 / 2
# No MW or BW specified or used
MW[0] = 0
BW[0] = 0
MWspec[0] = 0
BWspec[0] = 0

## Run model
def rotatein(ind, var, length):
    if ind <= length:
        retval = np.zeros(length)
        retval[:ind] = var[ind - 1 :: -1]
        return retval
    else:
        return var[ind - 1 : ind - length - 1 : -1]


history = 24
for ind in range(1, N):
    # Controller inputs
    Linput = rotatein(ind, L[1:], history)
    AMWinput = rotatein(ind, AMW, history)
    ABWinput = rotatein(ind, ABW, history)
    HUinput = rotatein(ind, HU, history)
    GUinput = rotatein(ind, GU, history)

    # Controller action
    MWspec[ind], BWspec[ind] = Controller(Linput, AMWinput, ABWinput, HUinput, GUinput)

    L[ind + 1], MW[ind], BW[ind], HU[ind], GU[ind], OF[ind] = TanksMassBalance(
        L[ind],
        RW[ind],
        AMW[ind],
        ABW[ind],
        MWspec[ind],
        BWspec[ind],
        HUreq[ind],
        GUreq[ind],
    )

# Trim level vector
L = L[:-1]

## Assess performance
cMW = 25
cBW = 60
cHU = 150
cGU = 70
J = (cMW * MW + cBW * BW + cHU * (HUreq - HU) + cGU * (GUreq - GU)).sum()

## Plots
# Disturbances
plt.figure()
plt.subplot(5, 1, 1)
plt.plot(RW)
plt.ylabel("RW")
plt.axis("tight")
plt.subplot(5, 1, 2)
plt.plot(HUreq)
plt.ylabel("HU_{req}")
plt.axis("tight")
plt.subplot(5, 1, 3)
plt.plot(GUreq)
plt.ylabel("GU_{req}")
plt.axis("tight")
plt.subplot(5, 1, 4)
plt.plot(AMW)
plt.ylabel("AMW")
plt.axis("tight")
plt.subplot(5, 1, 5)
plt.plot(ABW)
plt.ylabel("ABW")
plt.axis("tight")
plt.xlabel("Time step")

# Control results
plt.figure()
plt.subplot(5, 1, 1)
plt.plot(L)
plt.ylabel("L")
plt.axis("tight")
plt.subplot(5, 1, 2)
plt.plot(MW)
plt.ylabel("MW")
plt.axis("tight")
plt.subplot(5, 1, 3)
plt.plot(BW)
plt.ylabel("BW")
plt.axis("tight")
plt.subplot(5, 1, 4)
plt.plot(HUreq - HU)
plt.ylabel("\Delta HU")
plt.axis("tight")
plt.subplot(5, 1, 5)
plt.plot(GUreq - GU)
plt.ylabel("\Delta GU")
plt.axis("tight")
plt.xlabel("Time step")

print(J)
plt.show()

File: MATLAB - Contestants/ExampleSimulation.m
%% Load example, pre-allocate vectors
load Example1;
N = 365*24;
L = zeros(N+1,1);
MWspec = zeros(N,1);
BWspec = zeros(N,1);
MW = zeros(N,1);
BW = zeros(N,1);
HU = zeros(N,1);
GU = zeros(N,1);
OF = zeros(N,1);

%% Initialize model
% Tank is halfway full
L(1:2) = 2.155/2;
% No MW or BW specified or used
MW(1) = 0;
BW(1) = 0;
MWspec(1) = 0;
BWspec(1) = 0;

%% Run model
for ind = 2:N
    % Controller inputs
    % Level
    if ind < 24
        Linput = [L(ind:-1:1); zeros(24-ind,1)];
    else
        Linput = L(ind:-1:(ind-23));
    end
    % Usage and availability
    if ind < 25
        AMWinput = [AMW((ind-1):-1:1); zeros(24-ind,1)];
        ABWinput = [ABW((ind-1):-1:1); zeros(24-ind,1)];
        HUinput = [HU((ind-1):-1:1); zeros(24-ind,1)];
        GUinput = [GU((ind-1):-1:1); zeros(24-ind,1)];
    else
        AMWinput = AMW((ind-1):-1:(ind-24));
        ABWinput = ABW((ind-1):-1:(ind-24));
        HUinput = HU((ind-1):-1:(ind-24));
        GUinput = GU((ind-1):-1:(ind-24));        
    end
    
    % Controller action
    [MWspec(ind),BWspec(ind)] = Controller(Linput,AMWinput,ABWinput,HUinput,GUinput);
    
    % Mass balance
    [L(ind+1),MW(ind),BW(ind),HU(ind),GU(ind),OF(ind)] ...
    = TanksMassBalance(L(ind),RW(ind),AMW(ind),ABW(ind),...
    MWspec(ind),BWspec(ind),HUreq(ind),GUreq(ind));
end
% Trim level vector
L(end) = [];

%% Assess performance
cMW = 25;
cBW = 60;
cHU = 150;
cGU = 70;
J = sum(cMW*MW + cBW*BW + cHU*(HUreq-HU) + cGU*(GUreq-GU));

%% Plots
% Disturbances
figure;
subplot(5,1,1), plot(RW)
ylabel('RW')
axis tight
subplot(5,1,2), plot(HUreq)
ylabel('HU_r_e_q')
axis tight
subplot(5,1,3), plot(GUreq)
ylabel('GU_r_e_q')
axis tight
subplot(5,1,4), plot(AMW)
ylabel('AMW')
axis tight
subplot(5,1,5), plot(ABW)
ylabel('ABW')
axis tight
xlabel('Time step')

% Control results
figure;
subplot(5,1,1), plot(L)
ylabel('L')
axis tight
subplot(5,1,2), plot(MW)
ylabel('MW')
axis tight
subplot(5,1,3), plot(BW)
ylabel('BW')
axis tight
subplot(5,1,4), plot(HUreq-HU)
ylabel('\DeltaHU')
axis tight
subplot(5,1,5), plot(GUreq-GU)
ylabel('\DeltaGU')
axis tight
xlabel('Time step')
